src/engine/Distinct.cpp

```diff
@@ -36,17 +33,107 @@ VariableToColumnMap Distinct::computeVariableToColumnMap() const {
+  }
+}
+
 // _____________________________________________________________________________
-ProtoResult Distinct::computeResult([[maybe_unused]] bool requestLaziness) {
-  IdTable idTable{getExecutionContext()->getAllocator()};
+ProtoResult Distinct::computeResult(bool requestLaziness) {
   LOG(DEBUG) << "Getting sub-result for distinct result computation..." << endl;
-  std::shared_ptr<const Result> subRes = subtree_->getResult();
+  std::shared_ptr<const Result> subRes = subtree_->getResult(true);
 
   LOG(DEBUG) << "Distinct result computation..." << endl;
-  idTable.setNumColumns(subRes->idTable().numColumns());
-  size_t width = subRes->idTable().numColumns();
-  CALL_FIXED_SIZE(width, &Engine::distinct, subRes->idTable(), _keepIndices,
-                  &idTable);
-  LOG(DEBUG) << "Distinct result computation done." << endl;
-  return {std::move(idTable), resultSortedOn(), subRes->getSharedLocalVocab()};
+  size_t width = subtree_->getResultWidth();
+  if (subRes->isFullyMaterialized()) {
+    IdTable idTable =
+        CALL_FIXED_SIZE(width, &Distinct::distinct, subRes->idTable().clone(),
+                        _keepIndices, std::nullopt);
+    LOG(DEBUG) << "Distinct result computation done." << endl;
+    return {std::move(idTable), resultSortedOn(),
+            subRes->getSharedLocalVocab()};
+  }
+
+  auto generator = CALL_FIXED_SIZE(
+      width, &Distinct::lazyDistinct, std::move(subRes->idTables()),
+      _keepIndices,
+      requestLaziness ? std::nullopt
+                      : std::optional{IdTable{width, allocator()}});
+  if (!requestLaziness) {
+    IdTable result = cppcoro::getSingleElement(std::move(generator));
+    return {std::move(result), resultSortedOn(), subRes->getSharedLocalVocab()};
+  }
+  return {std::move(generator), resultSortedOn(),
+          subRes->getSharedLocalVocab()};
+}
+
+// _____________________________________________________________________________
+template <size_t WIDTH>
+IdTable Distinct::distinct(
+    IdTable dynInput, const std::vector<ColumnIndex>& keepIndices,
+    std::optional<typename IdTableStatic<WIDTH>::row_type> previousRow) {
+  AD_CONTRACT_CHECK(keepIndices.size() <= dynInput.numColumns());
+  LOG(DEBUG) << "Distinct on " << dynInput.size() << " elements.\n";
+  IdTableStatic<WIDTH> result = std::move(dynInput).toStatic<WIDTH>();
+
+  auto matchesRow = [&keepIndices](const auto& a, const auto& b) {
+    for (ColumnIndex i : keepIndices) {
+      if (a[i] != b[i]) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  // Variant of `std::ranges::unique` that allows to skip the first rows of
+  // elements found in the previous table.
+  auto first = std::ranges::find_if(result, [&matchesRow,
+                                             &previousRow](const auto& row) {
+    return !previousRow.has_value() || !matchesRow(row, previousRow.value());
+  });
+  auto last = result.end();
+
+  auto dest = result.begin();
+  if (first == dest) {
```

Comment:
As I see it,there are two optimizations:
1. Do this columnwise (see the Engine.cpp for a variation of this algorithm that only returns the count of unique elements)
2. Use `std::ranges::unique` (with your matches row you can build something)

3. What about cancellation? (see 1.)

Reply:
In total I think there might be a version with less code possible here.

Reply:
1. This would work, but it would require `O(numRows())` bytes of additional memory (the current algorithm is in-place, no additional memory required, we're just freeing memory that has been allocated by the previous operation).
2. After implementing this I'm fairly confident `std::ranges::unique` (as in this exact function) doesn't help here whatsoever, I really tried using it, but it just doesn't work for this use-case (with `previousRow` considered), but of course there might be other algorithms that might help making this shorter.
3. Yeah, cancellation would be nice, but that should be added once we settled on a general approach. For the non-lazy approach admittedly it would be better to not clone the whole thing at all and just copy the entries that are actually required, but this would require even more code.